src/processor/operator/aggregate/aggregate_hash_table.cpp

```diff
@@ -579,48 +562,61 @@ void AggregateHashTable::findHashSlots(const FactorizedTable& srcTable, uint64_t
+    std::vector<ValueVector*> distinctKeyVectors(keyVectors);
+    distinctKeyVectors.push_back(aggregateVector);
+    // The aggregateVector's state is either the same as the leading state (doesn't matter which we
+    // use), flat (where we must pass the original leading state), or unflat while the leadingState
+    // is flat (where we must pass the aggregateVector's state)
+    auto distinctLeadingState =
+        aggregateVector->state->isFlat() ? leadingState : aggregateVector->state.get();
+    append(distinctKeyVectors, distinctLeadingState, std::vector<AggregateInput>{},
+        1 /*multiplicity*/);
+}
+
+void AggregateHashTable::updateAggState(const std::vector<ValueVector*>& keyVectors,
+    AggregateFunction& aggregateFunction, ValueVector* aggVector, uint64_t multiplicity,
+    uint32_t aggStateOffset, const DataChunkState* leadingState) {
+    //  There may be a mix of flat and unflat states, but any unflat states will be the same
+    bool allFlat = leadingState->isFlat();
     if (!aggVector) {
-        updateNullAggVectorState(flatKeyVectors, unFlatKeyVectors, aggregateFunction, multiplicity,
-            aggStateOffset);
-    } else if (aggVector->state->isFlat() && unFlatKeyVectors.empty()) {
-        updateBothFlatAggVectorState(flatKeyVectors, aggregateFunction, aggVector, multiplicity,
-            aggStateOffset);
+        updateNullAggVectorState(*leadingState, aggregateFunction, multiplicity, aggStateOffset);
+    } else if (aggVector->state->isFlat() && allFlat) {
+        updateBothFlatAggVectorState(aggregateFunction, aggVector, multiplicity, aggStateOffset);
     } else if (aggVector->state->isFlat()) {
-        updateFlatUnFlatKeyFlatAggVectorState(flatKeyVectors, unFlatKeyVectors, aggregateFunction,
-            aggVector, multiplicity, aggStateOffset);
-    } else if (unFlatKeyVectors.empty()) {
-        updateFlatKeyUnFlatAggVectorState(flatKeyVectors, aggregateFunction, aggVector,
+        updateFlatUnFlatKeyFlatAggVectorState(*leadingState, aggregateFunction, aggVector,
             multiplicity, aggStateOffset);
-    } else if (!unFlatKeyVectors.empty() && (aggVector->state == unFlatKeyVectors[0]->state)) {
-        updateBothUnFlatSameDCAggVectorState(flatKeyVectors, unFlatKeyVectors, aggregateFunction,
-            aggVector, multiplicity, aggStateOffset);
+    } else if (allFlat) {
+        updateFlatKeyUnFlatAggVectorState(keyVectors, aggregateFunction, aggVector, multiplicity,
+            aggStateOffset);
+    } else if (aggVector->state.get() == leadingState) {
+        updateBothUnFlatSameDCAggVectorState(aggregateFunction, aggVector, multiplicity,
+            aggStateOffset);
     } else {
-        updateBothUnFlatDifferentDCAggVectorState(flatKeyVectors, unFlatKeyVectors,
-            aggregateFunction, aggVector, multiplicity, aggStateOffset);
+        updateBothUnFlatDifferentDCAggVectorState(*leadingState, aggregateFunction, aggVector,
```

Comment:
Can you add a test case for this line?

Reply:
It's covered now. I noticed that I'd forgotten to make the flattening of aggregates with a different group from the leadingState's group only apply to distinct aggregates. That was making this case impossible to hit.